src/utils/ScreenReader.test.ts

```diff
@@ -0,0 +1,353 @@
+        h('span', { 'aria-hidden': 'true' }, text('Nothing to see here'))
+      )
+    ).toBe(undefined)
+  })
+
+  it('does not describe elements hidden by display:none', () => {
+    const tts = fakeTTS()
+    const asr = new AriaScreenReader(tts)
+
+    expect(
+      asr.describe(
+        h('span', { style: 'display: none' }, text('Nothing to see here'))
+      )
+    ).toBe(undefined)
+  })
+
+  it('does not describe elements hidden by visibility:hidden', () => {
+    const tts = fakeTTS()
+    const asr = new AriaScreenReader(tts)
+
+    expect(
+      asr.describe(
+        h('span', { style: 'visibility: hidden' }, text('Nothing to see here'))
+      )
+    ).toBe(undefined)
+  })
+
+  it('does not describe document fragments', () => {
+    const tts = fakeTTS()
+    const asr = new AriaScreenReader(tts)
+
+    expect(asr.describe(document.createDocumentFragment())).toBe(undefined)
+  })
+
+  it('describes event targets if present', () => {
+    const tts = fakeTTS()
+    const asr = new AriaScreenReader(tts)
+
+    asr.speakEventTarget(h('button', text('Next')))
+    expect(tts.speak).toHaveBeenCalledWith('Next', expect.anything())
+  })
+
+  it('speaks event targets immediately rather than queueing speech', () => {
+    const tts = fakeTTS()
+    const asr = new AriaScreenReader(tts)
+
+    asr.speakEventTarget(h('button', text('Next')))
+    expect(tts.speak).toHaveBeenCalledWith(expect.anything(), { now: true })
+  })
+
+  it('does nothing if there is no event target', () => {
+    const tts = fakeTTS()
+    const asr = new AriaScreenReader(tts)
+
+    asr.speakEventTarget(undefined)
+    expect(tts.speak).not.toHaveBeenCalled()
+  })
+
+  it('speaks focus event targets immediately', () => {
+    const tts = fakeTTS()
+    const asr = new AriaScreenReader(tts)
+
+    asr.onFocus(h('button', text('Next')))
+    expect(tts.speak).toHaveBeenCalledWith('Next', { now: true })
+  })
+
+  it('speaks click event targets immediately', () => {
+    const tts = fakeTTS()
+    const asr = new AriaScreenReader(tts)
+
+    asr.onClick(h('button', text('Next')))
+    expect(tts.speak).toHaveBeenCalledWith('Next', { now: true })
+  })
+})
+
+describe('SpeechSynthesisTextToSpeech', () => {
+  it('speaks an utterance when given text to speak', () => {
+    const tts = new SpeechSynthesisTextToSpeech()
+
+    tts.speak('hello queued')
+    expect(speechSynthesis.cancel).not.toHaveBeenCalled()
+    expect(speechSynthesis.speak).toHaveBeenCalledWith(
+      expect.objectContaining({ text: 'hello queued' })
+    )
+  })
+
+  it('cancels speech before speaking an utterance if `now` is true', () => {
+    const tts = new SpeechSynthesisTextToSpeech()
+
+    tts.speak('hello right now!', { now: true })
+    expect(speechSynthesis.cancel).toHaveBeenCalled()
+    expect(speechSynthesis.speak).toHaveBeenCalledWith(
+      expect.objectContaining({ text: 'hello right now!' })
+    )
+  })
+
+  it('resolves when onend is called', async () => {
```

Comment:
This test ended up being tricky to write. Any suggestions for making it clearer/simpler would be appreciated.